File: openmetadata-sdk/src/main/java/org/openmetadata/sdk/entities/SearchIndex.java
package org.openmetadata.sdk.entities;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import org.openmetadata.schema.api.data.CreateSearchIndex;
import org.openmetadata.sdk.client.OpenMetadata;
import org.openmetadata.sdk.client.OpenMetadataClient;
import org.openmetadata.sdk.exceptions.OpenMetadataException;
import org.openmetadata.sdk.models.ListParams;
import org.openmetadata.sdk.models.ListResponse;

/**
 * SDK wrapper for SearchIndex operations.
 * This class provides static methods for SearchIndex CRUD operations.
 * It does NOT extend the schema SearchIndex class to avoid naming conflicts.
 */
public class SearchIndex {

  private static OpenMetadataClient defaultClient;

  public static void setDefaultClient(OpenMetadataClient client) {
    defaultClient = client;
  }

  private static OpenMetadataClient getClient() {
    if (defaultClient == null) {
      return OpenMetadata.client();
    }
    return defaultClient;
  }

  // Static methods for CRUD operations
  public static org.openmetadata.schema.entity.data.SearchIndex create(CreateSearchIndex request)
      throws OpenMetadataException {
    return getClient().searchIndexes().create(request);
  }

  public static org.openmetadata.schema.entity.data.SearchIndex retrieve(String id)
      throws OpenMetadataException {
    return getClient().searchIndexes().get(id);
  }

  public static org.openmetadata.schema.entity.data.SearchIndex retrieve(String id, String fields)
      throws OpenMetadataException {
    return getClient().searchIndexes().get(id, fields);
  }

  public static org.openmetadata.schema.entity.data.SearchIndex retrieve(UUID id)
      throws OpenMetadataException {
    return retrieve(id.toString());
  }

  public static org.openmetadata.schema.entity.data.SearchIndex retrieveByName(String name)
      throws OpenMetadataException {
    return getClient().searchIndexes().getByName(name);
  }

  public static org.openmetadata.schema.entity.data.SearchIndex retrieveByName(
      String name, String fields) throws OpenMetadataException {
    return getClient().searchIndexes().getByName(name, fields);
  }

  public static org.openmetadata.schema.entity.data.SearchIndex update(
      String id, org.openmetadata.schema.entity.data.SearchIndex patch)
      throws OpenMetadataException {
    return getClient().searchIndexes().update(id, patch);
  }

  public static void delete(String id) throws OpenMetadataException {
    getClient().searchIndexes().delete(id);
  }

  public static void delete(UUID id) throws OpenMetadataException {
    getClient().searchIndexes().delete(id);
  }

  public static void delete(String id, boolean recursive, boolean hardDelete)
      throws OpenMetadataException {
    Map<String, String> params = new HashMap<>();
    params.put("recursive", String.valueOf(recursive));
    params.put("hardDelete", String.valueOf(hardDelete));
    getClient().searchIndexes().delete(id, params);
  }

  // Async delete methods
  public static CompletableFuture<Void> deleteAsync(String id) {
    return getClient().searchIndexes().deleteAsync(id);
  }

  public static CompletableFuture<Void> deleteAsync(UUID id) {
    return getClient().searchIndexes().deleteAsync(id);
  }

  public static CompletableFuture<Void> deleteAsync(
      String id, boolean recursive, boolean hardDelete) {
    return CompletableFuture.runAsync(
        () -> {
          try {
            delete(id, recursive, hardDelete);
          } catch (OpenMetadataException e) {
            throw new RuntimeException(e);
          }
        });
  }

  public static ListResponse<org.openmetadata.schema.entity.data.SearchIndex> list()
      throws OpenMetadataException {
    return getClient().searchIndexes().list();
  }

  public static ListResponse<org.openmetadata.schema.entity.data.SearchIndex> list(
      ListParams params) throws OpenMetadataException {
    return getClient().searchIndexes().list(params);
  }
}

File: ingestion/src/metadata/sdk/entities/searchindex.py
"""
Search Index entity implementation for OpenMetadata Python SDK.
"""
from typing import List, Optional

from metadata.generated.schema.api.data.createSearchIndex import (
    CreateSearchIndexRequest,
)
from metadata.generated.schema.entity.data.searchIndex import (
    SearchIndex as SearchIndexEntity,
)
from metadata.sdk.entities.base import BaseEntity


class SearchIndex(BaseEntity):
    """Search Index entity operations for OpenMetadata SDK"""

    @classmethod
    def entity_type(cls):
        """Return the entity type"""
        return SearchIndexEntity

    @classmethod
    def create(cls, request: CreateSearchIndexRequest) -> SearchIndexEntity:
        """
        Create a new search index.

        Args:
            request: CreateSearchIndexRequest with search index details

        Returns:
            Created SearchIndex entity
        """
        client = cls._get_client()
        return client.create_or_update(request)

    @classmethod
    def retrieve(
        cls, entity_id: str, fields: Optional[List[str]] = None
    ) -> SearchIndexEntity:
        """
        Retrieve a search index by ID.

        Args:
            entity_id: SearchIndex UUID
            fields: Optional list of fields to include

        Returns:
            SearchIndex entity
        """
        client = cls._get_client()
        return client.get_by_id(
            entity=SearchIndexEntity, entity_id=entity_id, fields=fields
        )

    @classmethod
    def retrieve_by_name(
        cls, fqn: str, fields: Optional[List[str]] = None
    ) -> SearchIndexEntity:
        """
        Retrieve a search index by fully qualified name.

        Args:
            fqn: Fully qualified name
            fields: Optional list of fields to include

        Returns:
            SearchIndex entity
        """
        client = cls._get_client()
        return client.get_by_name(entity=SearchIndexEntity, fqn=fqn, fields=fields)

    @classmethod
    def update(cls, entity_id: str, entity: SearchIndexEntity) -> SearchIndexEntity:
        """
        Update a search index (PUT operation).

        Args:
            entity_id: SearchIndex UUID
            entity: Updated SearchIndex entity

        Returns:
            Updated SearchIndex entity
        """
        entity.id = entity_id
        client = cls._get_client()
        return client.create_or_update(entity)

    @classmethod
    def patch(cls, entity_id: str, json_patch: List[dict]) -> SearchIndexEntity:
        """
        Patch a search index (PATCH operation).

        Args:
            entity_id: SearchIndex UUID
            json_patch: JSON patch operations

        Returns:
            Patched SearchIndex entity
        """
        client = cls._get_client()
        return client.patch(
            entity=SearchIndexEntity, entity_id=entity_id, json_patch=json_patch
        )

    @classmethod
    def delete(
        cls, entity_id: str, recursive: bool = False, hard_delete: bool = False
    ) -> None:
        """
        Delete a search index.

        Args:
            entity_id: SearchIndex UUID
            recursive: Delete recursively
            hard_delete: Hard delete (permanent)
        """
        client = cls._get_client()
        client.delete(
            entity=SearchIndexEntity,
            entity_id=entity_id,
            recursive=recursive,
            hard_delete=hard_delete,
        )

    @classmethod
    def list(
        cls,
        fields: Optional[List[str]] = None,
        after: Optional[str] = None,
        before: Optional[str] = None,
        limit: int = 100,
    ) -> List[SearchIndexEntity]:
        """
        List search indexs.

        Args:
            fields: Fields to include
            after: Pagination cursor
            before: Pagination cursor
            limit: Number of results

        Returns:
            List of SearchIndex entities
        """
        client = cls._get_client()
        response = client.list_entities(
            entity=SearchIndexEntity,
            fields=fields,
            after=after,
            before=before,
            limit=limit,
        )
        return response.entities if response else []
